### ISSUE TITLE
SDK uses a mix of integers and floating point variables to represent "power"

### ISSUE THREAD

**Issue by stefan-brus-frequenz** [frequenz-floss/frequenz-sdk-python#28]
### What happened?

int: https://github.com/frequenz-floss/frequenz-sdk-python/blob/v0.x.x/src/frequenz/sdk/power_distribution/utils.py#L31
float: https://github.com/frequenz-floss/frequenz-sdk-python/blob/v0.x.x/src/frequenz/sdk/microgrid/component_data.py#L137

Different numeric types are used to represent power in this library.

### What did you expect instead?

One consistent type used across the entire library.

### Affected part(s)

Build script, CI, dependencies, etc. (part:tooling), Battery power distribution (part:power-distribution)

### Extra information

This forces the user to use manual type conversions, for example when charging a battery based on its own power bounds.

**Comment by leandro-lucarella-frequenz** [comment]
The SDK is following the microgrid API on this one and the microgrid API have this inconsistency because devices have these differences.

We agreed with @tiyash-basu-frequenz that anything that is device-dependent should be abstracted away by the microgrid API, so the microgrid will use `float` everywhere (which is the most flexible option) and will document that when setting power or other commands won't necessarily have decimal resolution (or even units resolution, as some devices have resolutions higher than 1w for example).

In any case, we can solve the inconsistency in the SDK now, by just doing the conversion here.

**Comment by daniel-zullo-frequenz** [comment]
Probably the root source of this issue is the microgrid API defines generic float bounds (see [here](https://github.com/frequenz-floss/frequenz-api-microgrid/blob/v0.x.x/proto/frequenz/api/microgrid/common.proto#L13-L21)) which are used to set the power bounds and any other metric bounds. However the microgrid API requires an integer to set the power (see [here](https://github.com/frequenz-floss/frequenz-api-microgrid/blob/v0.x.x/proto/frequenz/api/microgrid/microgrid.proto#L352-L362)

I have checked for power distributing and it is actually consistent with the microgrid API.

Anyway I'll push a draft PR to convert any power value to float and do the conversion to integer to set the power in the microgrid API so that we can discuss the changes but I still have my doubts this will improve the current state of the power distributing package and actor.

**Comment by tiyash-basu-frequenz** [comment]
Note that the API has a recent commit that accepts `float` as the power param type:

https://github.com/frequenz-floss/frequenz-api-microgrid/blob/9aa802daf4e042e3f7633309e09f0ff8df8f13fe/proto/frequenz/api/microgrid/microgrid.proto#L154-L158

https://github.com/frequenz-floss/frequenz-api-microgrid/blob/9aa802daf4e042e3f7633309e09f0ff8df8f13fe/proto/frequenz/api/microgrid/microgrid.proto#L366-L373

This is done using a new RPC called `SetPowerActive`, and the older `Charge` and `Discharge` methods have been deprecated.

**Comment by leandro-lucarella-frequenz** [comment]
> Probably the root source of this issue is the microgrid API defines generic float bounds (see [here](https://github.com/frequenz-floss/frequenz-api-microgrid/blob/v0.x.x/proto/frequenz/api/microgrid/common.proto?rgh-link-date=2023-04-13T15%3A22%3A30Z#L13-L21)) which are used to set the power bounds and any other metric bounds. However the microgrid API requires an integer to set the power (see [here](https://github.com/frequenz-floss/frequenz-api-microgrid/blob/v0.x.x/proto/frequenz/api/microgrid/microgrid.proto?rgh-link-date=2023-04-13T15%3A22%3A30Z#L352-L362)

Yes, this was mentioned in this issue :D

* https://github.com/frequenz-floss/frequenz-sdk-python/issues/28#issuecomment-1278804317

> Note that the API has a recent commit that accepts float as the power param type:

Ah, nice to see it is already there. I guess we need to coordinate the update of the SDK and do it only when the microgrid service is updated in the controller, right @tiyash-basu-frequenz ?

@daniel-zullo-frequenz the intention of this issue was to hide this from the user, and basically take `float` everywhere and convert to `int` to send to the microgrid API when needed. If it is such a trivial change as it sounds, I would go for it. If not maybe it is better to wait to update to the new microgrid API versions and just use the new calls that use `float`. But it really sounds like we can do all the work to update the SDK API to use `float`, it will be needed anyway when updating to the new microgrid API version.

**Comment by daniel-zullo-frequenz** [comment]
> Yes, this was mentioned in this issue :D

I believe I have provided more precise low level technical information aligned with what it was already described, probably for my own understanding. Anyway I'm happy you both jumped in to clarify it even more and to update me about the latest/coming changes in the microgrid API.

> the intention of this issue was to hide this from the user, and basically take float everywhere and convert to int to send to the microgrid API when needed. If it is such a trivial change as it sounds, I would go for it.

Yes, I totally understood that and I already worked out the changes, though I still need to fix all the unit tests failing due to float value comparison and my idea about the draft PR was mainly to show you and discuss if functions like `floor()`, `ceil()` should be kept or not as they made a difference and probably they will be no longer needed having only float values. It is not entirely trivial as we need to be careful in a few places where there are arithmetic operations involved. Anyway this is the direction it needs to go based on Tiyash's comments.

**Comment by tiyash-basu-frequenz** [comment]
>Ah, nice to see it is already there. I guess we need to coordinate the update of the SDK and do it only when the microgrid service is updated in the controller, right

Yes, that's right :)
I will, of course, inform the SDK team once the microgrid service is ready.
